src/MediatR/NotificationPublishers/TaskWhenAllPublisher.cs

```diff
@@ -0,0 +1,28 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace MediatR.NotificationPublishers;
+
+/// <summary>
+/// Uses Task.WhenAll with the list of Handler tasks:
+/// <code>
+/// var tasks = handlers
+///                .Select(handler => handler.Handle(notification, cancellationToken))
+///                .ToList();
+/// 
+/// return Task.WhenAll(tasks);
+/// </code>
+/// </summary>
+public class TaskWhenAllPublisher : INotificationPublisher
+{
+    public Task Publish(IEnumerable<NotificationHandlerExecutor> handlerExecutors, INotification notification, CancellationToken cancellationToken)
+    {
+        var tasks = handlerExecutors
+            .Select(handler => handler.HandlerCallback(notification, cancellationToken))
+            .ToList();
```

Comment:
tiny nit: using '.ToArray()' hits the most optimized path of Task.WhenAll. 

(passing an IEnumerable/ICollection to Task.WhenAll internally creates another array of tasks, discarding the given list, passing a Task array directly or using the params overload does not, saving an allocation)